src/libxdata/world/models/airport/Airport.cpp

```diff
@@ -150,14 +171,51 @@ const std::shared_ptr<Runway> Airport::getRunwayByName(const std::string& rw) co
     return rwy->second;
 }
 
-void Airport::forEachRunway(std::function<void(const std::shared_ptr<Runway>)> f) {
+void Airport::forEachRunway(std::function<void(const std::shared_ptr<Runway>)> f) const {
     for (auto &rwy: runways) {
         f(rwy.second);
     }
 }
 
+void Airport::forEachRunwayPair(std::function<void(const std::shared_ptr<Runway>, const std::shared_ptr<Runway>)> f) const {
+    for (auto &rwys: runwayPairs) {
+        f(rwys.first, rwys.second);
+    }
+}
+
 bool Airport::hasOnlyHeliports() const {
-    return runways.empty();
+    return runways.empty() && !heliports.empty();
+}
+
+bool Airport::hasWaterRunway() const {
+    for (const auto &rwy: runways) {
+        if (rwy.second->getSurfaceType() == xdata::Runway::SurfaceType::WATER_RUNWAY)
+            return true;
+    }
+    return false;
+}
+
+bool Airport::hasTowerFrequency() const {
+    int numberOfTowerFrequencies = 0;
+    try {
+        const auto towerFrequencies = atcFrequencies.at(ATCFrequency::TWR);

```

Comment:
Using `at` and expecting an exception is not a good idea, this will slow down huge iterations a lot because exceptions are rather expensive. This could be rewritten using `map::find`, comparing to `atcFrequencies.end()`.